internal/sys/fd_trace.go

```diff
@@ -0,0 +1,137 @@
+package sys
+
+import (
+	"bytes"
+	"fmt"
+	"runtime"
+	"strings"
+	"sync"
+)
+
+// FDs is a registry of all file descriptors wrapped into sys.FDs that were
+// created while an fd tracer was active.
+var FDs = newFDMap()
+
+// Finalize is called when the Go runtime frees an [FD].
+var Finalize FDTracer
+
+// traceFDs returns true if there are active fd tracers registered to the
+// package.
+func traceFDs() bool {
+	return Finalize != nil
+}
+
+type fdmap struct {
```

Comment:
This should be replaced with a https://pkg.go.dev/sync#Map instead, to save on code we have to maintain.

Reply:
sync.Map doesn't have the equivalent of len() and iterating is quite clunky. I'll give it a try after addressing the other feedback.

Reply:
Went with sync.Map after all.